gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/MessageConsumerImpl.java

```diff
@@ -0,0 +1,306 @@
+import com.google.cloud.pubsub.spi.PubSubRpc.PullCallback;
+import com.google.cloud.pubsub.spi.PubSubRpc.PullFuture;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+
+import io.grpc.internal.SharedResourceHolder;
+
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Default implementation for a message consumer.
+ */
+final class MessageConsumerImpl implements MessageConsumer {
+
+  private static final int MAX_QUEUED_CALLBACKS = 100;
+  // shared scheduled executor, used to schedule pulls
+  private static final SharedResourceHolder.Resource<ScheduledExecutorService> TIMER =
+      new SharedResourceHolder.Resource<ScheduledExecutorService>() {
+        @Override
+        public ScheduledExecutorService create() {
+          ScheduledThreadPoolExecutor timer = new ScheduledThreadPoolExecutor(1);
+          timer.setRemoveOnCancelPolicy(true);
+          return timer;
+        }
+
+        @Override
+        public void close(ScheduledExecutorService instance) {
+          instance.shutdown();
+        }
+      };
+
+  private final PubSubOptions pubsubOptions;
+  private final PubSubRpc pubsubRpc;
+  private final PubSub pubsub;
+  private final AckDeadlineRenewer deadlineRenewer;
+  private final String subscription;
+  private final MessageProcessor messageProcessor;
+  private final ScheduledExecutorService timer;
+  private final ExecutorFactory<ExecutorService> executorFactory;
+  private final ExecutorService executor;
+  private final AtomicInteger queuedCallbacks;
+  private final int maxQueuedCallbacks;
+  private final Object futureLock = new Object();
+  private final Runnable consumerRunnable;
+  private final RestartPolicy restartPolicy;
+  private boolean closed;
+  private Future<?> scheduledFuture;
+  private PullFuture pullerFuture;
+
+  /**
+   * Interface for policies according to which the consumer should be restarted.
+   */
+  interface RestartPolicy {
+
+    boolean shouldRestart(int queuedCallbacks);
+  }
+
+  /**
+   * Default restart policy. Restarts the consumer once {@code restartThreshold} messages out of
+   * {@code maxQueuedCallbacks} have already been processed.
+   */
+  static class DefaultRestartPolicy implements RestartPolicy {
+
+    final int maxQueuedCallbacks;
+    final int restartThreshold;
+
+    DefaultRestartPolicy(int maxQueuedCallbacks, int restartThreshold) {
+      this.maxQueuedCallbacks = maxQueuedCallbacks;
+      this.restartThreshold = restartThreshold;
+    }
+
+    @Override
+    public boolean shouldRestart(int queuedCallbacks) {
+      return (maxQueuedCallbacks - queuedCallbacks) >= restartThreshold;
```

Comment:
Why `<=`? We should restart once more or equals to `restartThreshold` callbacks have completed.